various fixes for nwnltr.c

Lord-Nightmare opened the pull request:
Fixes the idx off-by-one error for ' and -
Fixes LF characters on stdin on windows causing idx to return -1 and index off the bottom of the letter probability arrays
Fixes input names shorter than 3 characters causing an out of bounds array access and crash
Adds a more verbose display of invalid input characters in case they are not printed in a distinguishable way
Flush stderr after writing to it, since otherwise in some circumstances it doesn't get flushed to console at a consistent point in execution
Add a compile time definition of the number of letters in the file, so both 26 and 28 letter variants can be handled

Lord-Nightmare committed:
Fix off-by-one error with the ' and - characters in the idx() function
Prevent scanf() from overrunning the buf[256] array when reading an input line
Fix an array out-of-bounds access during list building when an input name is less than 3 characters long
Add ability to handle standard input on windows/mingw (with LF at the end of each line)

Lord-Nightmare committed:
Flush stderr after writing to it, to make output more consistent in certain environments

Lord-Nightmare committed:
nwnltr.c: Made the number of letters into a #define, so both 26 and 28 letter variant files can be handled, chosen at compile time
Added a hex display for invalid input characters, so unprintable characters can be distinguished.

Lord-Nightmare committed:
nwnltr.c: Stop parsing an input string on encountering a # character, to allow unparsed comments

Lord-Nightmare committed:
Fix backtracking during name generation so that it cannot backtrack past the 4th letter, this should prevent generating bogus 3-letter names.

mtijanic commented on nwnltr.c:
```diff
@@ -46,6 +46,7 @@ struct cfg {
 #define die(format, ...)                                \
     do {                                                \
         fprintf(stderr, format "\n", ##__VA_ARGS__);    \
+        fflush(stderr);                                 \
```
This is super weird, since stderr is supposed to be line-buffered, which means it's always flushed on `\n`. Doesn't hurt though.

Lord-Nightmare replied:
The way the later versions (since about 5 years ago) of mingw32/msys2 on windows implement stdout/stderr/stdin is a little screwy (ends up with gdb's exit prompt always auto-pressing Y and a few other weird things going on). I don't think its strictly necessary at all, but it doesn't seem to hurt. I can get rid of it, if needed.

mtijanic commented on nwnltr.c:
```diff
@@ -80,20 +81,21 @@ void parse_cmdline(int argc, char *argv[]) {
 // Files can have fewer (just alpha), but there is no point as the special ones
 // can just be given a probability of 0 to achieve the same effect.
 // Thus, 28 is hardcoded here, for ease of file IO.
+#define MAX_LETTERS 28
```
Good call. I'd in fact do 
```C
#ifndef MAX_LETTERS
#define MAX_LETTERS 28
#endif
```

so you could override it from command line with `-D`. Maybe `s/MAX_LETTERS/NUM_LETTERS/g`

Lord-Nightmare replied:
Good idea. I'll do that.

mtijanic commented on nwnltr.c:
```diff
@@ -128,21 +130,24 @@ void load_ltr(const char *filename, struct ltrfile *ltr) {
-    while (scanf("%s", buf) == 1) {
+    while (scanf("%255s", buf) == 1) {
         for (q = buf; *q; q++) {
+            if (((*q) == 0x0d) || ((*q) == 0x0a) || ((*q) == '#')) // stop on CR, LF, or #
```
How does this ever return LF? Or CR for that matter? `%s` should scan everything that's not whitespace.

Lord-Nightmare replied:
On mingw32/msys2, it returns a string like "arthur\n\0", I've never actually seen it return a CR but it definitely returns the LF before the NUL. Why? I don't know. EDIT: I'm actually having trouble reproducing this now, but it isn't relevant anymore, see below.

This actually indicates another bug, to be honest: if an invalid character is found during a string parse, it prints a warning about skipping that character to stderr, but doesn't actually skip it at all! It still ends up in the buf[] string, and idx[p[*]] will return -1 making it index 'off the bottom' of the various probability arrays. I'm not sure how best to deal with this, sort of constructing an entirely new 'sanitized' string in a buf2[] array each iteration of the while() loop.
Fixing this would remove the need to explicitly halt parsing on any character (except '#' indicating a comment) as any invalid character would just be ignored completely and not copied to buf2[]
EDIT2: I've added exactly this mechanism, see below.

mtijanic reviewed (approved):
If you want I can merge this right now and we can look at the other stuff in a follow up.

Lord-Nightmare committed:
Rename MAX_LETTERS macro to NUM_LETTERS for consistency, and allow it to be overridden by the compiler command line

Lord-Nightmare committed:
Use a separate array to sanitize input names after they are read in
Made error messages on skipping a character or name more verbose
Made character pointers local to the loops in which they get used

Lord-Nightmare commented:
I added an extra buffer where only the sanitized characters from the input string end up, to prevent any issues with indexing out of bounds if a character is invalid. I also got rid of the explicit checking for CR/LF on input (only leaving the check for '#' to immediately terminate parsing a comment) since the new sanitizer mechanism will catch these and any other input garbage and throw them out.

Lord-Nightmare commented:
If you want to cherry-pick/merge certain subparts, go ahead. I'm away until mid-weekend, though.

mtijanic commented:
Merging the whole lot, and any other stuff can come as a follow-up.
